### PR TITLE
Make the worker role optional

### PR DESCRIPTION
**Pull request by tillrohrmann**
This commit changes the Node implementation to make the worker role optional. This also includes the node-ctrl service which can run a worker grpc service if the worker role is configured. If a node does not run the cluster controller role locally, then one needs to configure it with a cluster_controller_address.

### CODE REVIEW

File: crates/node/src/lib.rs [hunk 1]
```diff
@@ -70,56 +73,61 @@ pub enum BuildError {
...
+            options.cluster_controller_address
+        {
+            if cluster_controller_role.is_some() {
+                warn!("This node is running the cluster controller but has also a remote cluster controller address configured. \
+                This indicates a potential misconfiguration. Trying to connect to the remote cluster controller.");
+            }
+
+            cluster_controller_address
+        } else if cluster_controller_role.is_some() {
+            NetworkAddress::DnsName(format!("localhost:{}", node_ctrl.port()))
```
Comment (AhmedSoliman) [inline crates/node/src/lib.rs:123]:
```suggestion
            NetworkAddress::DnsName(format!("127.0.0.1:{}", node_ctrl.port()))
```

Nit: Avoid ns lookups whenever possible.

File: crates/node/src/lib.rs [hunk 2]
```diff
@@ -188,13 +201,18 @@ impl Node {
...
 
-    async fn attach_node(node_id: NodeId, cluster_controller_endpoint: Uri) -> Result<(), Error> {
-        info!("Attach to cluster at '{cluster_controller_endpoint}'");
-        let channel = Channel::builder(cluster_controller_endpoint.clone())
-            .connect_timeout(Duration::from_secs(5))
-            .connect_lazy();
+    async fn attach_node(
+        node_id: PlainNodeId,
+        cluster_controller_address: NetworkAddress,
+    ) -> Result<(), Error> {
```
Comment (AhmedSoliman) [inline crates/node/src/lib.rs:207]: Will this return the NodeId in the future?

File: crates/node/src/lib.rs [hunk 3]
```diff
@@ -205,32 +223,46 @@ impl Node {
...
-            ClusterControllerLocation::Local
-        } else {
-            ClusterControllerLocation::Remote(s.to_string())
+    fn create_channel_from_network_address(
+        cluster_controller_address: &NetworkAddress,
+    ) -> Result<Channel, http::Error> {
+        let channel = match cluster_controller_address {
+            NetworkAddress::Uds(uds_path) => {
+                let uds_path = uds_path.clone();
+                Endpoint::try_from("/")
```
Comment (AhmedSoliman) [inline crates/node/src/lib.rs:237]: Worth leaving a comment explaining that it's a dummy endpoint.

File: crates/node/src/lib.rs [hunk 4]
```diff
@@ -205,32 +223,46 @@ impl Node {
...
+            .scheme("http")
+            .authority(authority.to_string())
+            .path_and_query("/")
+            .build()
+    }
 
-        Ok(result)
+    fn create_lazy_channel_from_uri(uri: Uri) -> Channel {
+        Channel::builder(uri)
+            .connect_timeout(Duration::from_secs(5))
```
Comment (AhmedSoliman) [inline crates/node/src/lib.rs:265]: Hopefully is is configurable in the future.

File: crates/node/src/lib.rs [hunk 5]
```diff
@@ -205,32 +223,46 @@ impl Node {
...
+                let uri = Self::create_uri(dns_name)?;
+                Self::create_lazy_channel_from_uri(uri)
+            }
         };
+        Ok(channel)
+    }
+
+    fn create_uri(authority: impl ToString) -> Result<Uri, http::Error> {
+        Uri::builder()
+            .scheme("http")
```
Comment (AhmedSoliman) [inline crates/node/src/lib.rs:257]: A todo that we need to make this configurable?

File: crates/types/src/nodes_config.rs [hunk 6]
```diff
@@ -101,14 +104,17 @@ impl NodeConfig {
     }
 }
 
-#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+#[derive(Debug, Clone, Eq, PartialEq, Hash, derive_more::Display)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
 pub enum NetworkAddress {
     /// Unix domain socket
+    #[display(fmt = "unix:{}", _0)]
```
Comment (AhmedSoliman) [inline crates/types/src/nodes_config.rs:111]: 🤝

**Review by AhmedSoliman — APPROVED** [review]
Nice! A couple of nits/questions but this is good to go.

File: crates/node/src/lib.rs (hunk 2 above)
Comment (tillrohrmann) [reply crates/node/src/lib.rs:207]: Yes, this is the plan.

File: crates/node/src/lib.rs (hunk 3 above)
Comment (tillrohrmann) [reply crates/node/src/lib.rs:237]: 👍

File: crates/node/src/lib.rs (hunk 4 above)
Comment (tillrohrmann) [reply crates/node/src/lib.rs:265]: Yes. I will add a comment expressing the intention.

File: crates/node/src/lib.rs (hunk 5 above)
Comment (tillrohrmann) [reply crates/node/src/lib.rs:257]: Yes, will add it.

**Pull request merged**